contracts/cw-ics721-bridge/src/ibc_packet_receive.rs

```diff
@@ -1,40 +1,48 @@
 use cosmwasm_std::{
-    from_binary, to_binary, Addr, DepsMut, Empty, Env, IbcPacket, IbcReceiveResponse, StdResult,
-    SubMsg, WasmMsg,
+    from_binary, to_binary, Addr, Binary, DepsMut, Empty, Env, IbcPacket, IbcReceiveResponse,
+    StdResult, SubMsg, WasmMsg,
 };
 
+use zip_optional::Zippable;
+
 use crate::{
     ibc::{NonFungibleTokenPacketData, ACK_AND_DO_NOTHING},
     ibc_helpers::{get_endpoint_prefix, try_pop_source_prefix},
-    msg::{CallbackMsg, ExecuteMsg, NewTokenInfo, TransferInfo},
+    msg::{CallbackMsg, ExecuteMsg},
     state::{INCOMING_CLASS_TOKEN_TO_CHANNEL, OUTGOING_CLASS_TOKEN_TO_CHANNEL, PO},
+    token_types::{Class, ClassId, Token, TokenId, VoucherCreation, VoucherRedemption},
     ContractError,
 };
 
 /// Every incoming token has some associated action.
 enum Action {
-    /// We have seen this token before, it should be transfered.
-    Transfer { class_id: String, token_id: String },
-    /// We have not seen this token before, a new one needs to be
-    /// created.
-    NewToken {
-        class_id: String,
-        token_id: String,
-        token_uri: String,
+    /// Debt-voucher redemption.
+    Redemption {
```

Comment:
Can we get a little explanation on what this new struct is? How does it work etc? (On this PR, no need to clutter up code with comments)